src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -68,17 +68,37 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mmods_try_matched_recv(void *buf,
 #ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
     /* XPMEM special receive */
     if (MPIDI_SHM_REQUEST(message, status) & MPIDI_SHM_REQ_XPMEM_SEND_LMT) {
+        MPIDI_XPMEM_am_unexp_rreq_t *unexp_rreq = &MPIDI_XPMEM_REQUEST(message, unexp_rreq);
+        MPIR_Comm *root_comm = NULL;
+        int recvtype_iscontig;
+
+        MPIR_Datatype_is_contig(datatype, &recvtype_iscontig);
+        if (recvtype_iscontig) {
+            MPIDI_XPMEM_REQUEST(message, counter) =
+                &MPIDI_XPMEM_global.coop_counter[MPIDI_XPMEM_global.local_rank *
+                                                 MPIDI_XPMEM_global.num_local +
+                                                 unexp_rreq->src_lrank];
+
+            root_comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(message, context_id));
+
+            mpi_errno = MPIDI_XPMEM_lmt_cts_isend(buf, count, datatype,
+                                                  MPIDIG_REQUEST(message, rank), root_comm,
+                                                  &message);
+            MPIR_ERR_CHECK(mpi_errno);
+        } else {
+            MPIDI_XPMEM_REQUEST(message, counter) = NULL;
+        }
+
```

Comment:
How data is copied if `recvtype_iscontig` is FALSE?

Reply:
it falls back to XPMEM single copy